The new chisel (updated in Rocket on 02/03/2017) does not work with riscv-boom.

Old version (works): e27657118ff5915b96f8e3a467d464245fe09769
New version (does not work): 97871178cb511063965f971b768f91c289c4776f

When compiling the riscv-boom code, an exception occurs in this line:
https://github.com/freechipsproject/chisel3/blob/97871178cb511063965f971b768f91c289c4776f/chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala#L617

```
error text:
...
[error] Caused by: chisel3.core.AutoClonetypeException: Unable to automatically infer cloneType on class boom.IssueSlotIO$$anon$1: Unable to determine instance of outer class class boom.IssueSlotIO, multiple possible candidates List(boom.IssueSlotIO@69a5c, boom.IssueSlotIO@69da4) assignable to outer class type
[error] 	at chisel3.core.Bundle.chisel3$core$Bundle$$autoClonetypeError$1(Aggregate.scala:576)
[error] 	at chisel3.core.Bundle$$anonfun$11.apply(Aggregate.scala:617)
[error] 	at chisel3.core.Bundle$$anonfun$11.apply(Aggregate.scala:593)
[error] 	at scala.Option.map(Option.scala:146)
[error] 	at chisel3.core.Bundle.cloneType(Aggregate.scala:593)
[error] 	at chisel3.core.Bundle.cloneType(Aggregate.scala:511)
[error] 	at chisel3.core.Output$.apply(Data.scala:180)
[error] 	at Chisel.package$AddDirectionToData$.asOutput$extension(compatibility.scala:28)
[error] 	at boom.IssueSlotIO.<init>(issue_slot.scala:42)
[error] 	at boom.IssueSlotIO.cloneType(issue_slot.scala:45)
[error] 	at boom.IssueSlotIO.cloneType(issue_slot.scala:19)
...
```
the riscv-boom code in which the error occurs:
```scala
...
19: class IssueSlotIO(num_wakeup_ports: Int)(implicit p: Parameters) extends BoomBundle()(p)
20: {
...
36:   val debug = {
37:     val result = new Bundle {
38:       val p1 = Bool()
39:       val p2 = Bool()
40:       val p3 = Bool()
41:    }
42:    result.asOutput
43:  }
44:
45:   override def cloneType = new IssueSlotIO(num_wakeup_ports)(p).asInstanceOf[this.type]
...
```

What do you think about it?

--- a/chiselFrontend/src/main/scala/chisel3/internal/Builder.scala
+++ b/chiselFrontend/src/main/scala/chisel3/internal/Builder.scala
@@ -173,8 +173,8 @@ private[chisel3] class DynamicContext() {
   var currentClockAndReset: Option[ClockAndReset] = None
   val errors = new ErrorLog
   val namingStack = new internal.naming.NamingStack
-  // Record the Bundle instance, class name, and reverse stack trace position of open Bundles
-  val bundleStack: ArrayBuffer[(Bundle, String, Int)] = ArrayBuffer()
+  // Record the Bundle instance, class name, method name, and reverse stack trace position of open Bundles
+  val bundleStack: ArrayBuffer[(Bundle, String, String, Int)] = ArrayBuffer()
 }
 
 private[chisel3] object Builder {
@@ -247,13 +247,19 @@ private[chisel3] object Builder {
   // Returns the current stack of open Bundles
   // Note: elt will NOT have finished construction, its elements cannot be accessed
   def updateBundleStack(elt: Bundle): Seq[Bundle] = {
-    val stackClasses = Thread.currentThread().getStackTrace()
-        .map(_.getClassName)
+    val stackElts = Thread.currentThread().getStackTrace()
         .reverse  // so stack frame numbers are deterministic across calls
+        .dropRight(2)  // discard Thread.getStackTrace and updateBundleStack
+
+    // Determine where we are in the Bundle stack
+    val eltClassName = elt.getClass.getName
+    val eltStackPos = stackElts.map(_.getClassName).lastIndexOf(eltClassName)
 
     // Prune the existing Bundle stack of closed Bundles
-    val pruneLength = dynamicContext.bundleStack.reverse.prefixLength { case (_, cname, pos) =>
-      pos >= stackClasses.size || stackClasses(pos) != cname
+    // If we know where we are in the stack, discard frames above that
+    val stackEltsTop = if (eltStackPos >= 0) eltStackPos else stackElts.size
+    val pruneLength = dynamicContext.bundleStack.reverse.prefixLength { case (_, cname, mname, pos) =>
+      pos >= stackEltsTop || stackElts(pos).getClassName != cname || stackElts(pos).getMethodName != mname
     }
     dynamicContext.bundleStack.trimEnd(pruneLength)
 
@@ -261,10 +267,9 @@ private[chisel3] object Builder {
     val lastStack = dynamicContext.bundleStack.map(_._1).toSeq
 
     // Append the current Bundle to the stack, if it's on the stack trace
-    val eltClassName = elt.getClass.getName
-    val eltStackPos = stackClasses.lastIndexOf(eltClassName)
     if (eltStackPos >= 0) {
-      dynamicContext.bundleStack.append((elt, eltClassName, eltStackPos))
+      val stackElt = stackElts(eltStackPos)
+      dynamicContext.bundleStack.append((elt, eltClassName, stackElt.getMethodName, eltStackPos))
     }
     // Otherwise discard the stack frame, this shouldn't fail noisily
 

--- a/src/test/scala/chiselTests/AutoClonetypeSpec.scala
+++ b/src/test/scala/chiselTests/AutoClonetypeSpec.scala
@@ -56,6 +56,12 @@ object CompanionObjectWithBundle {
   }
 }
 
+class NestedAnonymousBundle extends Bundle {
+  val a = Output(new Bundle {
+    val a = UInt(8.W)
+  })
+}
+
 // A Bundle with an argument that is also a field.
 // Not necessarily good style (and not necessarily recommended), but allowed to preserve compatibility.
 class BundleWithArgumentField(val x: Data, val y: Data) extends Bundle
@@ -146,4 +152,13 @@ class AutoClonetypeSpec extends ChiselFlatSpec {
       io.data := 1.U
     } }
   }
+
+  "Nested directioned anonymous Bundles" should "not need clonetype" in {
+    elaborate { new Module {
+      val io = IO(new NestedAnonymousBundle)
+      val a = WireInit(io)
+      io.a.a := 1.U
+    } }
+  }
+
 }

